frontend/dockerfile/dockerfile2llb/directives.go

```diff
@@ -53,3 +95,70 @@ func ParseDirectives(r io.Reader) map[string]Directive {
 	}
 	return m
 }
+
+func commentDirectiveParser(comment string) directiveParser {
+	return &regexpDirectiveParser{
+		src: fmt.Sprintf(`^%s\s*([a-zA-Z][a-zA-Z0-9]*)\s*=\s*(.+?)\s*$`, comment),
+	}
+}
+
+type jsonDirectiveParser struct{}
+
+func (p *jsonDirectiveParser) parse(r io.Reader) map[string]Directive {
+	data := map[string]interface{}{}
+	dec := json.NewDecoder(r)
+	dec.Decode(&data)
+	if data == nil {
+		return nil
+	}
+
+	v, ok := data[keySyntax]
+	if !ok {
+		return nil
+	}
+	vs, ok := v.(string)
+	if !ok {
+		return nil
+	}
+
+	return map[string]Directive{
+		keySyntax: {
+			Name:  keySyntax,
+			Value: vs,
+		},
+	}
+}
+
+type allowShebangParser struct {
+	parser directiveParser
+}
+
+func allowShebang(parser directiveParser) directiveParser {
+	return &allowShebangParser{
+		parser: parser,
+	}
+}
+
+func (p *allowShebangParser) parse(r io.Reader) map[string]Directive {
+	lr := bufio.NewReader(r)
+	line, err := lr.ReadString('\n')
+	if err != nil {
+		return nil
+	}
+
+	if !strings.HasPrefix(line, "#!") {
+		// no shebang, parse as normal
+		return p.parser.parse(io.MultiReader(bytes.NewReader([]byte(line)), lr))
+	}
+
+	directives := p.parser.parse(lr)
+	for k, v := range directives {
+		for i, loc := range v.Location {
```

Comment:
This isn't really needed as shebang is not valid in Dockerfile (it breaks the rules of Dockerfile directives). It only comes from the generalization that Dockerfile directives and syntax from non-Dockerfiles need to be collected with the same function.